nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/autodiff/samediff/SameDiff.java

```diff
@@ -3198,48 +3512,120 @@ public void createGradFunction() {
                     } else {
                         inputsToOp = sameDiff.ops.get(df.getOwnName()).getInputsToOp();
                         outputsOfOp = sameDiff.ops.get(df.getOwnName()).getOutputsOfOp();
-                        numProcessed++;
                     }
 
 
                     //Get gradients for all output variables:
                     List<SDVariable> grads = new ArrayList<>();
                     for(String s : outputsOfOp){
-                        SDVariable g = sameDiff.getVariable(s).gradient();
-                        Preconditions.checkNotNull(g, "Could not get gradient for variable %s as output of op %s", g.getVarName(), df.getOwnName());
-                        grads.add(g);
+                        SDVariable v = sameDiff.getVariable(s);
+                        SDVariable g = v.hasGradient() ? v.gradient() : null;
+
+                        if(g == null){
+                            //If no gradient exists at this point, 3 possibilities:
+                            // (a) we have a bug
+                            // (b) output of this op isn't used in calculating the loss
+                            // (c) output isn't a FP type
+                            //In the FP case, we should create a zero variable to backprop, because we can't perform backprop
+                            // for this op otherwise...
+                            if(!v.dataType().isFPType()){
+                                grads.add(null);
+                            } else {
+                                SDVariable gTemp = sameDiff.zerosLike(v);
```

Comment:
Should this even happen? Calculating with a zero gradient, should basically be "don't do any change at all" and therefore we shouldn't be performing backprop at all in that case?

Reply:
This is a little unintuitive, but it's definitely intended (and required).
Consider the following graph:
```
input -> split -> (A,B)
B -> lossFn
```
Here, A is unused (in that it doesn't contribute to the loss function). But to do split op backprop, we need gradient variables/arrays for both outputs (A and B).
We know the shape and type of dL/dA must be exactly the same as A; we also know that the loss function doesn't depend on A. Hence, dL/dA == zerosLike(A)

Reply:
got it. Maybe add the explanation to the comment on top?

Reply:
Done.